src/mpi/coll/transports/generic/transport.c

```diff
@@ -0,0 +1,1243 @@
+                   sizeof(int) * old_vtcs[i].invtcs.used);
+            memcpy(sched->vtcs[i].outvtcs.array, old_vtcs[i].outvtcs.array,
+                   sizeof(int) * old_vtcs[i].outvtcs.used);
+            /* free the memory allocated by the old vertex */
+            MPIR_COLL_GENERIC_free_vtx(&old_vtcs[i]);
+        }
+        /* free the arary of old vertices */
+        MPIR_COLL_GENERIC_free_mem(old_vtcs);
+    }
+    /* return last unused vertex */
+    *vtx = &sched->vtcs[sched->total];
+    return sched->total++;      /* return vertex id */
+}
+
+/* Transport function that adds a no op vertex in the graph that has
+ * all the vertices posted before it as incoming vertices */
+int MPIR_COLL_GENERIC_fence(MPIR_COLL_GENERIC_sched_t * sched)
+{
+    MPL_DBG_MSG_FMT(MPIR_DBG_COLL,VERBOSE,(MPL_DBG_FDEST,"TSP(generic) : sched [fence] total=%d \n", sched->total));
+
+    MPIR_COLL_GENERIC_vtx_t *vtxp;
+    int *invtcs, i, n_invtcs = 0, vtx_id;
+
+    vtx_id = MPIR_COLL_GENERIC_get_new_vtx(sched, &vtxp);
+
+    vtxp->kind = MPIR_COLL_GENERIC_KIND_FENCE;
+    MPIR_COLL_GENERIC_init_vtx(sched, vtxp, vtx_id);
+
+    invtcs = (int *) MPIR_COLL_GENERIC_allocate_mem(sizeof(int) * vtx_id);
+    /* record incoming vertices */
+    for (i = vtx_id - 1; i >= 0; i--) {
+        if (sched->vtcs[i].kind == MPIR_COLL_GENERIC_KIND_WAIT)
+            /* no need to record this and any vertex before wait.
+             * Dependency on the last wait call will be added by
+             * the subsequent call to MPIR_COLL_GENERIC_add_vtx_dependencies function */
+            break;
+        else {
+            invtcs[vtx_id - 1 - i] = i;
+            n_invtcs++;
+        }
+    }
+
+    MPIR_COLL_GENERIC_add_vtx_dependencies(sched, vtx_id, n_invtcs, invtcs);
+    MPL_free(invtcs);
+    return vtx_id;
+}
+
+/* Transport function that adds a no op vertex in the graph with specified dependencies */
+int MPIR_COLL_GENERIC_wait_for(MPIR_COLL_GENERIC_sched_t * sched, int nvtcs, int *vtcs)
+{
+    MPL_DBG_MSG_FMT(MPIR_DBG_COLL,VERBOSE,(MPL_DBG_FDEST,"TSP(generic) : sched [wait_for] total=%d \n", sched->total));
+    MPIR_COLL_GENERIC_vtx_t *vtxp;
+    int vtx_id = MPIR_COLL_GENERIC_get_new_vtx(sched, &vtxp);
+
+    vtxp->kind = MPIR_COLL_GENERIC_KIND_NOOP;
+    MPIR_COLL_GENERIC_init_vtx(sched, vtxp, vtx_id);
+
+    MPIR_COLL_GENERIC_add_vtx_dependencies(sched, vtx_id, nvtcs, vtcs);
+    return vtx_id;
+}
+
+/* MPIR_COLL_GENERIC_wait waits for all the operations posted before it to complete
+* before issuing any operations posted after it. This is useful in composing
+* multiple schedules, for example, allreduce can be written as
+* MPIR_ALGO_sched_reduce(s)
+* MPIR_TSP_wait(s)
+* MPIR_ALGO_sched_bcast(s)
+* This is different from the fence operation in the sense that fence requires
+* a vertex to post dependencies on it while MPIR_COLL_GENERIC_wait is used internally
+* by the transport to add it as a dependency to any operations poster after it
```

Comment:
"poster" -> "posted"

Reply:
The comment above and the code below doesn't make it clear to me how this is different from fence.

Reply:
I will put an example and make it more clear.